bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/events/OSGiEventManager.java

```diff
@@ -0,0 +1,229 @@
+                }
+            }
+        }
+    }
+
+    protected void removeEventSubscriber(EventSubscriber eventSubscriber) {
+        Set<String> subscribedEventTypes = eventSubscriber.getSubscribedEventTypes();
+
+        for (String subscribedEventType : subscribedEventTypes) {
+            typedEventSubscribers.remove(subscribedEventType, eventSubscriber);
+        }
+    }
+
+    @Override
+    public void handleEvent(org.osgi.service.event.Event osgiEvent) {
+        Object typeObj = osgiEvent.getProperty("type");
+        Object payloadObj = osgiEvent.getProperty("payload");
+        Object topicObj = osgiEvent.getProperty("topic");
+        Object sourceObj = osgiEvent.getProperty("source");
+
+        if (typeObj instanceof String && payloadObj instanceof String && topicObj instanceof String) {
+            String typeStr = (String) typeObj;
+            String payloadStr = (String) payloadObj;
+            String topicStr = (String) topicObj;
+            String sourceStr = (sourceObj instanceof String) ? (String) sourceObj : null;
+            if (!typeStr.isEmpty() && !payloadStr.isEmpty() && !topicStr.isEmpty()) {
+                handleEvent(typeStr, payloadStr, topicStr, sourceStr);
+            }
+        } else {
+            logger.error("The handled OSGi event is invalid. Expect properties as string named 'type', 'payload' and 'topic'. "
+                    + "Received event properties are: " + osgiEvent.getPropertyNames());
+        }
+    }
+
+    private void handleEvent(final String type, final String payload, final String topic, final String source) {
+        EventFactory eventFactory = typedEventFactories.get(type);
+
+        if (eventFactory != null) {
+            Set<EventSubscriber> eventSubscribers = getEventSubscribers(type);
+            if (!eventSubscribers.isEmpty()) {
+                Event eshEvent = createESHEvent(eventFactory, type, payload, topic, source);
+                if (eshEvent != null) {
+                    dispatchESHEvent(eventSubscribers, eshEvent);
+                }
+            }
+        } else {
+            logger.warn("Could not find an Event Factory for the event type '" + type + "'.");
+        }
+    }
+
+    private Event createESHEvent(final EventFactory eventFactory, final String type, final String payload,
+            final String topic, final String source) {
+        Event eshEvent = null;
+        try {
+            eshEvent = eventFactory.createEvent(type, topic, payload, source);
+        } catch (Exception e) {
+            logger.error("Creation of ESH-Event failed, "
+                    + "because one of the registered event factories has thrown an exception: " + e.getMessage(), e);
+        }
+        return eshEvent;
+    }
+
+    private void dispatchESHEvent(final Set<EventSubscriber> eventSubscribers, final Event event) {
+        for (EventSubscriber eventSubscriber : eventSubscribers) {
```

Comment:
We really need to be aware that we took a step backwards here regarding robustness of the system, since a single subscriber can now completely block our system. We should really address https://bugs.eclipse.org/bugs/show_bug.cgi?id=423517 soon for removing this risk.